pallets/committee/src/lib.rs

```diff
@@ -289,76 +298,104 @@ decl_module! {
+                None => Err(Error::<T, I>::FirstVoteReject.into()),
+            }
         }
 
-        /// Enact the referendum
+        /// Votes `approve`ingly (or not, if `false`)
+        /// on an existing `proposal` given by its hash, `index`.
         ///
         /// # Arguments
-        /// * `id` - Pip Id that need to be enacted
-        #[weight = (T::DbWeight::get().reads_writes(6, 2) + 400_000_000, Operational, Pays::Yes)]
-        pub fn vote_enact_referendum(origin, id: PipId) -> DispatchResult {
-            Self::vote_referendum( origin, id,
-                || {
-                    let aye_call = T::EnactProposalMaker::enact_referendum_call(id);
-                    let nay_call = T::EnactProposalMaker::reject_referendum_call(id);
-                    (aye_call, nay_call)
-                })?;
-            let current_did = Context::current_identity::<Identity<T>>().unwrap_or_default();
-            Self::deposit_event(RawEvent::VoteEnactReferendum(current_did, id));
-            Ok(())
-        }
-
-        /// Reject the referendum
+        /// * `proposal` - A hash of the proposal to be voted on.
+        /// * `index` - The proposal index.
+        /// * `approve` - If `true` than this is a `for` vote, and `against` otherwise.
         ///
-        /// # Arguments
-        /// * `id` - Pip Id that need to be rejected
-        #[weight = (T::DbWeight::get().reads_writes(6, 2) + 400_000_000, Operational, Pays::Yes)]
-        pub fn vote_reject_referendum(origin, id: PipId) -> DispatchResult {
-            Self::vote_referendum( origin, id,
-                || {
-                    let aye_call = T::EnactProposalMaker::enact_referendum_call(id);
-                    let nay_call = T::EnactProposalMaker::reject_referendum_call(id);
-                    (nay_call, aye_call)
-                })?;
-            let current_did = Context::current_identity::<Identity<T>>().unwrap_or_default();
-            Self::deposit_event(RawEvent::VoteRejectReferendum(current_did, id));
+        /// # Errors
+        /// * `BadOrigin`, if the `origin` is not a member of this committee.
+        #[weight = (500_000_000, Operational, Pays::Yes)]
+        pub fn vote(
+            origin,
+            proposal: T::Hash,
+            index: ProposalIndex,
+            approve: bool,
+        ) -> DispatchResult {
+            // 1. Ensure `origin` is a committee member.
+            let did = Self::ensure_is_member(origin)?;
+
+            // 2. Ensure a prior proposal exists and that their indices match.
+            let mut voting = Self::voting(&proposal).ok_or(Error::<T, I>::NoSuchProposal)?;
+            ensure!(voting.index == index, Error::<T, I>::MismatchedVotingIndex);
+
+            // 3. Vote on aye / nay and remove from the other.
+            let aye = (voting.ayes.iter().position(|a| a == &did), &mut voting.ayes);
+            let nay = (voting.nays.iter().position(|a| a == &did), &mut voting.nays);
```

Comment:
Unbounded loops are never great. Can we refactor it store the required data in mapping so that we can have O(1) access?

Reply:
Checking the position should be quite cheap in terms of constant factor even if `O(n)` though, so assuming committees are reasonably sized (say <= 30), it seems like a pessimization / more complex to add 2 additional maps. But sure, we can do that if you prefer?

Reply:
I'll capture this as a separate story for discussion. We could bound the size of committees to some sensible number as an alternative.